### ISSUE TITLE
 Git credentials not set in container when using non-root user

### ISSUE THREAD

**Issue by olad32** [microsoft/vscode-remote-release#720]
- VSCode Version: 1.35.1
- Local OS Version: Win 10
- Remote OS Version: Debian Stretch
- Remote Extension/Connection Type: Docker

Steps to Reproduce:

1. Launch with non root user
.devcontainerconfig.json
```
{
  "name": "Node.js",
  "dockerFile": "Dockerfile-devenv",
  "appPort": [4000],
  "runArgs": ["--user", "node", "--env", "HOME=/home/node"],
  "postCreateCommand": "npm install"
}
```

Dockerfile-devenv
```
ARG NODE_VERSION=10.15.3-stretch
ARG NPM_VERSION=6.4.1

FROM node:${NODE_VERSION}

# Avoid warnings by switching to noninteractive
ENV DEBIAN_FRONTEND=noninteractive
# Set the default shell to bash instead of sh
ENV SHELL /bin/bash

# Configure apt and install packages
RUN apt-get update \
    && apt-get -y install --no-install-recommends apt-utils 2>&1 \ 
    #
    # Verify git and needed tools are installed
    && apt-get install -y git procps \
    #
    # Clean up
    && apt-get autoremove -y \
    && apt-get clean -y \
    && rm -rf /var/lib/apt/lists/* \
    #
    # Update npm
    && npm install npm@${NPM_VERSION}

# Switch back to dialog for any ad-hoc use of apt-get
ENV DEBIAN_FRONTEND=dialog
```

2. Everything works well excepted that i cannot "git commit", git complains : 
```
*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"
```

3. As root (ie remove runArgs from .devcontainer.json) i can "git commit" without git complaining

Extract from the Dev Container log when running in non-root user : 
`Run: docker exec -w /home/node/.vscode-server/bin/c7d83e57cd18f18026a8162d042843bda1bcf21f -u 0 b361fc82da06ce9192c081b6b0159e7a2ded1a9e454c7b52803bf66e3fd71dff /bin/sh -c command -v git >/dev/null 2>&1 && git config --system credential.helper '!f() { command -v code >/dev/null 2>&1 && code --gitCredential $*; }; f' || true`

Does this issue occur when you try this locally?: No
Does this issue occur when you try this locally and all extensions are disabled?: No

**Comment by chrmarti** [comment]
Could you try VS Code Insiders and confirm this is still a problem there?

**Comment by olad32** [comment]
Yes just tried with VS Code Insiders v1.36.0 (Extension "Remote Containers" v0.62.0), same problem

**Comment by olad32** [comment]
In addition i tried to "git push", but it did not work either as root or non-root user. 
Git ask me for my credentials to access my git repo although my credentials for this repo are effectively stored in Windows credential store.
My local git config for credential.helper is manager, does the remote extension is able to transfer these types of credentials (Windows store) to the container ?

I don't know if this is related to the first problem with git commit, but maybe it is the same root cause.

**Comment by chrmarti** [comment]
Could you try adding `ENV HOME /home/node` to the Dockerfile? We are not parsing `"runArgs"` for the HOME environment variable and misplace `.gitconfig` if we don't have that.

**Comment by olad32** [comment]
> Could you try adding `ENV HOME /home/node` to the Dockerfile? We are not parsing `"runArgs"` for the HOME environment variable and misplace `.gitconfig` if we don't have that.

Thanks this fix the first issue, i am now able to commit with non-root user

**Comment by olad32** [comment]
Do i open another issue for the second problem ? git push do not work as Windows git credentials do not seems to be applied in the container

**Comment by chrmarti** [comment]
Let's continue here. What do you get for `which code` in the terminal in VS Code when you are connected to the container?

**Comment by olad32** [comment]
As root user : 
root@1171823d9e2e:/workspaces/projectA# which code-insiders
/root/.vscode-server-insiders/bin/6520c22b83c0b6f168ba7de3309ffab42281e905/bin/code-insiders

**Comment by chrmarti** [comment]
Could you try rebuilding the container (use either 'Rebuild Container' command or `docker rm ...`) with VS Code Insiders? That should place it under `/home/node`.

**Comment by olad32** [comment]
Yes, as node user : 
/home/node/.vscode-server-insiders/bin/6520c22b83c0b6f168ba7de3309ffab42281e905/bin/code-insiders

**Comment by chrmarti** [comment]
What do you get for `git config --get credential.helper` as node user?

**Comment by olad32** [comment]
`node@2eb811bdb36c:/workspaces/projectA$ git config --get credential.helper
!f() { command -v code-insiders >/dev/null 2>&1 && code-insiders --gitCredential $*; }; f`

**Comment by olad32** [comment]
Here is an extract of the container launch log : 
```
Run: docker exec -w / -u 0 2eb811bdb36c2834b3eb86baa4ad369937a16a6be83d1a8bdc3659b1ecfd59ff /bin/sh -c command -v git >/dev/null 2>&1 && git config --system credential.helper '!f() { command -v code-insiders >/dev/null 2>&1 && code-insiders --gitCredential $*; }; f' || true
```

**Comment by chrmarti** [comment]
That looks all good. Can you double-check that `git push` works with your credentials on Windows?

**Comment by olad32** [comment]
Yes it works when i open the folder locally : 
```
fapl01831@PFIEFRET159 MINGW64 /c/sources/projectA (migration)
$ git pull
Already up to date.
```

As opposed to when i open the folder in container : 
```
node@2eb811bdb36c:/workspaces/projectA$ git pull
'git' n'est pas reconnu en tant que commande interne
ou externe, un programme exécutable ou un fichier de commandes.
Username for 'https://gitlab-repo-gpf.apps.eul.sncf.fr': 
```

**Comment by olad32** [comment]
I think i got it. I think this because i use git bash for windows, vscode tries to get credentials from classic cmd cli where git is not available. 
Is it fixable in vscode ?

**Comment by chrmarti** [comment]
@joaomoreno Does the Git extension support Git Bash?

**Comment by joaomoreno** [comment]
We only support Git for Windows, which comes with Git Bash. When VS Code invokes git.exe it doesn't use Git Bash at all.

**Comment by chrmarti** [comment]
I'll close this then. I would add support for Git Bash only after VS Code's Git extension has done so.
